agent/tailog/store.go

```diff
@@ -0,0 +1,100 @@
+// pmm-agent
+// Copyright 2019 Percona LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package tailog help to store tail logs
+package tailog
+
+import (
+	"container/ring"
+	"strings"
+	"sync"
+)
+
+// Store implement ring save logs.
+type Store struct {
+	log      *ring.Ring
+	capacity int
+	m        sync.Mutex
+}
+
+// NewStore creates Store.
+func NewStore(capacity int) *Store {
+	return &Store{
+		log:      ring.New(capacity),
+		capacity: capacity,
+	}
+}
+
+// Write writes log for store.
+func (l *Store) Write(b []byte) (int, error) {
+	l.m.Lock()
+	defer l.m.Unlock()
+
+	// when store 0 logs
+	if l.log == nil {
+		return len(b), nil
+	}
+
+	l.log.Value = string(b)
+	l.log = l.log.Next()
+	return len(b), nil
+}
+
+// Resize to update capacity.
+func (l *Store) Resize(capacity int) {
+	l.m.Lock()
+	defer l.m.Unlock()
+
+	if l.capacity == capacity {
+		return
+	}
+
+	old := l.log
```

Comment:
looks like we can use `ring.Link` and `ring.Unlink` methods here to increase and decrease size of ring

Reply:
I saw this methods before;
In our case `ring.Link` and `ring.Unlink` will complicate the code, because i need calculate it's increase or decrease and write additional test to check that all work fine;